kpatch-build/create-diff-object.c

```diff
@@ -3421,7 +3486,11 @@ static void kpatch_create_mcount_sections(struct kpatch_elf *kelf)
 			nr++;
 
 	/* create text/rela section pair */
+#ifdef __aarch64__
+	sec = create_section_pair(kelf, "__patchable_function_entries", sizeof(void*), nr);
```

Comment:
When trying to kpatch-build a change to meminfo_proc_show(), I got the following error:
```
make[1]: Entering directory '/root/linux'
  LDS     /root/.kpatch/tmp/patch/kpatch.lds
  CC [M]  /root/.kpatch/tmp/patch/patch-hook.o
  LD [M]  /root/.kpatch/tmp/patch/livepatch-a.o
ld: __patchable_function_entries has both ordered [`__patchable_function_entries' in /root/.kpatch/tmp/patch/patch-hook.o] and unordered [`__patchable_function_entries' in /root/.kpatch/tmp/patch/output.o] sections
ld: final link failed: bad value
make[2]: *** [scripts/Makefile.build:431: /root/.kpatch/tmp/patch/livepatch-a.o] Error 1
```
The problem is that the compiler-created version of the section has the `SHF_LINK_ORDER` flag set.  I "fixed" it with the following hack.  `sh_link` just points to a random section but maybe that's ok, AFAICT its value doesn't matter for the kernel's use of `__patchable_function_entries`.
```
diff --git a/kpatch-build/create-diff-object.c b/kpatch-build/create-diff-object.c
index 11a5f94..bd34764 100644
--- a/kpatch-build/create-diff-object.c
+++ b/kpatch-build/create-diff-object.c
@@ -3488,6 +3488,8 @@ static void kpatch_create_mcount_sections(struct kpatch_elf *kelf,
        /* create text/rela section pair */
 #ifdef __aarch64__
        sec = create_section_pair(kelf, "__patchable_function_entries", sizeof(void*), nr);
+       sec->sh.sh_flags |= SHF_LINK_ORDER;
+       sec->sh.sh_link = 1;
 #else /* !__aarch64__ */
        sec = create_section_pair(kelf, "__mcount_loc", sizeof(void*), nr);
 #endif
```

Reply:
Ok so I have the opposite problem in that patch-hook.o is unordered and output.o becomes ordered with the change you suggested which and the linking fails.

Which version of gcc are you using?

Reply:
I've seen this on both rhel9 and fedora34 (`gcc version 11.2.1 20210728 (Red Hat 11.2.1-1) (GCC)`). We might want to copy `sh_flags`/`sh_link` from `__patchable_function_entries` of patched objectfile.

Reply:
Yeah that sounds like the best option